### PR TITLE
[spark] Fix code for getting databricks entry_point, and remove useless code in databricks hook

### LINKED ISSUE #40223: Ray Databricks Hook Throws AttributeError at startup
### What happened + What you expected to happen

**Reproduction Steps**
Follow official documentation from Databricks: https://docs.databricks.com/en/machine-learning/ray-integration.html
When calling this code block:
```python
from ray.util.spark import setup_ray_cluster, shutdown_ray_cluster

setup_ray_cluster(
  num_worker_nodes=2,
  num_cpus_per_node=4,
  collect_log_to_path="/dbfs/path/to/ray_collected_logs"
)
```

An AttributeError is thrown (relevant traceback part)
```txt
    733 # wait background spark task starting.
    734 for _ in range(_BACKGROUND_JOB_STARTUP_WAIT):

File /local_disk0/.ephemeral_nfs/envs/pythonEnv-7df49a7d-a6a5-4d31-819f-3db0882f8a36/lib/python3.10/site-packages/ray/util/spark/databricks_hook.py:90, in DefaultDatabricksRayOnSparkStartHook.on_cluster_created(self, ray_cluster_handler)
     89 def on_cluster_created(self, ray_cluster_handler):
---> 90     db_api_entry = _get_db_api_entry()
     91     try:
     92         db_api_entry.registerBackgroundSparkJobGroup(
     93             ray_cluster_handler.spark_job_group_id
     94         )

File /local_disk0/.ephemeral_nfs/envs/pythonEnv-7df49a7d-a6a5-4d31-819f-3db0882f8a36/lib/python3.10/site-packages/ray/util/spark/databricks_hook.py:74, in _get_db_api_entry()
     70 def _get_db_api_entry():
     71     """
     72     Get databricks API entry point.
     73     """
---> 74     return get_dbutils().entry_point

AttributeError: 'DBUtils' object has no attribute 'entry_point'
```

### Versions / Dependencies

**Environment**

Using Azure Databricks, runtime: 13.3 LTS ML (includes Apache Spark 3.4.1, Scala 2.12)

Some python packages:
```txt
Requirement already satisfied: lightgbm_ray in /local_disk0/.ephemeral_nfs/envs/pythonEnv-639521b0-b362-4284-b2c9-1ba46ae35d88/lib/python3.9/site-packages (0.1.9)
Requirement already satisfied: deltalake in /local_disk0/.ephemeral_nfs/envs/pythonEnv-639521b0-b362-4284-b2c9-1ba46ae35d88/lib/python3.9/site-packages (0.11.0)
Requirement already satisfied: deltaray in /local_disk0/.ephemeral_nfs/envs/pythonEnv-639521b0-b362-4284-b2c9-1ba46ae35d88/lib/python3.9/site-packages (0.2.0)
Requirement already satisfied: lightgbm in /databricks/python3/lib/python3.9/site-packages (3.3.4)
Requirement already satisfied: packaging in /databricks/python3/lib/python3.9/site-packages (from lightgbm_ray) (21.3)
Requirement already satisfied: xgboost-ray>=0.1.12 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-639521b0-b362-4284-b2c9-1ba46ae35d88/lib/python3.9/site-packages (from lightgbm_ray) (0.1.19)
Requirement already satisfied: pyarrow<=12,>=8 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-639521b0-b362-4284-b2c9-1ba46ae35d88/lib/python3.9/site-packages (from deltalake) (12.0.0)
Requirement already satisfied: numpy>=1.24.1 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-639521b0-b362-4284-b2c9-1ba46ae35d88/lib/python3.9/site-packages (from deltaray) (1.26.0)
Requirement already satisfied: ray[data]>=2.2.0 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-639521b0-b362-4284-b2c9-1ba46ae35d88/lib/python3.9/site-packages (from deltaray) (2.7.1)
```

### Reproduction script

In Azure Databricks Runtime, execute:

```python
from ray.util.spark import setup_ray_cluster, shutdown_ray_cluster

setup_ray_cluster(
  num_worker_nodes=2,
  num_cpus_per_node=4,
  collect_log_to_path="/dbfs/path/to/ray_collected_logs"
)
```

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by WeichenXu123**
## Why are these changes needed?

Fix code for getting databricks entry_point, and remove useless code in databricks hook.

This removed code calls internal (but unstable) databricks API, but we don't need it actually. See explanation: https://github.com/ray-project/ray/pull/40221#issuecomment-1784124696

## Related issue number

Closes https://github.com/ray-project/ray/issues/40223
Closes https://github.com/ray-project/ray/pull/40221

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by paolorechia — APPROVED** [review]
Thanks for looking into this.

**Review by jjyao — APPROVED** [review]

**Comment by jjyao** [comment]
Lint failure:

```
Tue Oct 31 02:38:26 UTC 2023 Flake8....
--
  | python/ray/util/spark/databricks_hook.py:107:13: F821 undefined name 'db_api_entry'
  | python/ray/util/spark/databricks_hook.py:137:29: F821 undefined name 'db_api_entry'
  | 🚨 Error: The command exited with status 123


```

**Comment by WeichenXu123** [comment]
emm, `getIdleTimeMillisSinceLastNotebookExecution` still relies on entrypoint, let me check..

**Comment by jjyao** [comment]
@WeichenXu123 any progress on this PR?

**Comment by WeichenXu123** [comment]
> @WeichenXu123 any progress on this PR?

I will file a new PR :)

File: python/ray/util/spark/databricks_hook.py [hunk 1]
```diff
@@ -9,25 +9,15 @@
...
-        ip_shell = IPython.get_ipython()
-        if ip_shell is None:
-            raise _NoDbutilsError
-        return ip_shell.ns_table["user_global"]["dbutils"]
-    except ImportError:
-        raise _NoDbutilsError
-    except KeyError:
-        raise _NoDbutilsError
+    user_namespace_initializer = UserNamespaceInitializer.getOrCreate()
+    return user_namespace_initializer.get_spark_entry_point()
```
Comment (WeichenXu123) [inline python/ray/util/spark/databricks_hook.py:20]: Note:

I synced with databricks colleagues, we shouldn't use `dbutils.entry_point` or `dbutils.jobs.entry_point` to get entry_point, but we should use UserNamespaceInitializer instead.

**Review by jjyao — APPROVED** [review]

**Review by paolorechia — APPROVED** [review]

**Comment by jjyao** [comment]
Test failed: 

```
AttributeError: 'module' object at ray.util.spark.databricks_hook has no attribute '_get_db_api_entry'
```

**Comment by jjyao** [comment]
Test failed: https://buildkite.com/ray-project/premerge/builds/11442

**Review by jjyao — APPROVED** [review]

**Pull request merged**